### PR TITLE
fix(web): keep composer draft when stopping during a pending approval

### LINKED ISSUE #8963: [Bug]: Typed composer text is lost when clicking Stop during a pending approval prompt
Loving T3 Code, using it daily to drive agents on a second machine. Sharing this one in case it helps.

When the agent asked a multiple-choice question (pick 1/2/3), I didn't want any of the options, so I typed what I actually wanted into the composer. The choice card stayed pending, so I clicked Stop to interrupt, and everything I had typed was discarded. I had to retype the whole message.

Expected: composer drafts survive Stop (persist per thread), or typed text while a choice is pending gets submitted as a free-form answer.

Seen on the desktop app v0.0.37 on macOS, thread running against a remote SSH-launched environment.

Thanks for the great tool.

### PR DESCRIPTION
**Pull request by lnieuwenhuis**
While an agent's multiple-choice question is pending, the composer routes typed text into that question's custom-answer state instead of the persisted draft. When the question disappears without being answered, via Stop, a provider-side abort, or any other cancellation, that text was discarded, so a user who typed a free-form reply and then clicked Stop had to retype everything.

The fix watches the pending question transition itself rather than the Stop button: when the active request id changes and the request was not submitted, the typed custom answer is merged into the thread's composer draft. A non-blank existing draft is preserved with the rescued text appended after a blank line rather than overwritten. Rescue is gated on the same thread still being on screen, so navigating between threads while a question is open never writes into the wrong draft, and a failed submit does not prevent a later real Stop from rescuing the text.

The decision and merge logic are pure helpers in `ChatView.logic.ts` with tests for cancel, same-render question replacement, failed-submit-then-cancel, submitted, thread switch, and blank-answer cases. Mobile keeps the question's free-text field separate from its composer and is not affected.

Deliberately erasing a carried first-question answer also clears its matching persisted copy, so Stop does not resurrect text the user deleted. Unrelated drafts and option-selection behavior are preserved. No visual change; the composer keeps the text the user intended.

Closes #8963

Claude Fable 5.1 via Claude Code

Carried draft ownership now uses the stable environment/thread/request identity across navigation. The existing lifecycle lives in a focused hook used by ChatView, with mounted React tests using the real composer store. Regressions cover A → B → A erase/edit, failed responses after navigation, successful submissions, and cross-environment isolation. All 278 focused tests, web typecheck, and targeted format/lint checks passed. Full browser/router navigation was not exercised.

Review verification: a fresh web TypeScript check with a new build-info file and `--listFiles` passed and explicitly included the mounted test. The three `submission` reads occur inside later callbacks, so the reported TS2454 error does not occur; no definite-assignment assertions were added. The mounted tests reuse the existing installed renderer without changing dependencies. A future test-renderer migration is separate from this draft-ownership fix.

---

> [!NOTE]
> **Medium Risk**
> Changes composer draft and pending-user-input bookkeeping across Stop, submit, failure, and thread navigation; logic is mostly pure and tested, but regressions could still affect draft content in edge timing cases.
> 
> **Overview**
> Fixes **#8963**: composer text was lost when a pending agent question vanished without a submit (Stop, abort, teardown), because typing went into the question’s custom-answer state instead of the persisted draft.
> 
> **`ChatView.logic.ts`** adds pure helpers to decide when to rescue cancelled requests (same thread, not submitted), join multi-question custom answers, merge rescued text into the draft (append vs replace, no duplicates), carry an existing draft into a newly shown custom-answer question, and return answer-slot text to the draft when picking an option.
> 
> **`ChatView.tsx`** wires a layout effect on pending-request transitions to rescue unsent answers and seed new questions from the draft; tracks submitted request ids and “carried” draft copies so successful submits and thread switches do not corrupt the wrong draft. **`onRespondToUserInput`** marks submits early, clears carried draft on send, and on failure restores draft plus rescues if the question already disappeared. **`onSelectActivePendingUserInputOption`** moves typed custom text back to the draft before applying the selected option.
> 
> Unit tests cover rescue, carry, merge, and edge cases; no intended UI change beyond keeping the user’s text.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 027df29b19344d7c01625880db2cb531ff05a202. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Fix composer draft loss when pending user input is cancelled or replaced
> - When a pending user-input request disappears or is replaced on the same draft target, non-blank custom answers merge back into the composer draft.
> - Eligible non-blank composer text is carried into the first custom-answer state of a newly arrived request.
> - Selecting a pending-question option returns the existing typed custom answer to the composer before applying the option.
> - Adds utilities in [ChatView.logic.ts](https://github.com/pingdotgg/t3code/pull/9117/files#diff-464e876afd6be3800453dd8cc94d805066953bed62ffd825bbd6fdfd86f4d114) for answer aggregation, draft merging, and rescue eligibility.
> - Behavioral Change: `onRespondToUserInput` now tracks submitted requests; failed submissions can restore carried drafts and suppress thread errors for interrupted commands.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 027df29.</sup>
> 
>

## Summary by CodeRabbit

* **New Features**
  * Preserved composer drafts while custom-answer questions are pending.
  * Restored unsent answers when questions are cancelled or submission fails.
  * Combined existing drafts with nonblank custom answers to reduce retyping.
  * Retained draft text when selecting options or navigating between pending questions.
  * Preserved unrelated draft text during answer recovery and navigation.
  * Kept other carried answers and draft content when an answer is erased.

* **Tests**
  * Added coverage for draft preservation, recovery, cancellation, navigation, failures, and custom-answer handling.

### CODE REVIEW

**Review by Mnigos — CHANGES_REQUESTED** [review]
Tested this locally and the rescue itself works: typed a reply while a question was pending, hit Stop, my text came back into the draft. Switching threads keeps the typed answer, and a submitted one doesn't come back as a bogus rescue.

Honestly though, I don't think this fully fixes #8963. The moment the question card shows up, whatever you typed visibly vanishes from the composer and only comes back merged after Stop. Rescue makes the loss recoverable, but the text disappearing at all is the confusing part. I'd want the composer to just keep its content while a question is pending.

Side note: I couldn't reach the multi-question case the bots flagged through the UI at all, typing an answer turns the button into Next, so two unanswered custom answers at once don't seem reachable by hand.

File: apps/web/src/components/ChatView.tsx [hunk 1]
```diff
@@ -2419,6 +2432,55 @@ function ChatViewContent(props: ChatViewProps) {
...
+      // typed text is still readable here.
+      let answerForPrevious: string | null = null;
+      for (const draft of Object.values(
+        pendingUserInputAnswersByRequestId[previous.requestId] ?? {},
+      )) {
+        if (draft.customAnswer && draft.customAnswer.trim().length > 0) {
+          answerForPrevious = draft.customAnswer;
+          break;
+        }
+      }
```
[the remark this answers is not in the document]
Comment (lnieuwenhuis) [reply apps/web/src/components/ChatView.tsx]: Addressed in fdb8ecc4e: the rescue now joins every non-blank custom answer of the request via collectPendingUserInputCustomAnswers, with a test for the multi-question case.

File: apps/web/src/components/ChatView.logic.ts [hunk 2]
```diff
@@ -245,6 +245,73 @@ export function buildThreadTurnInterruptInput(thread: Pick<Thread, "id" | "sessi
...
+  wasSubmitted: boolean;
+}): string | null {
+  const { previous, nextRequestId, currentDraftTarget, answerForPrevious, wasSubmitted } = input;
+  if (!previous) {
+    return null;
+  }
+  if (previous.requestId === nextRequestId) {
+    return null;
+  }
+  if (wasSubmitted) {
```
[the remark this answers is not in the document]
Comment (lnieuwenhuis) [reply apps/web/src/components/ChatView.logic.ts]: Addressed in fdb8ecc4e: the submitted mark is removed when the RPC fails, and a failed submit whose question is already gone rescues the answer directly, so a later real Stop or the failure itself still recovers the text.

File: apps/web/src/components/ChatView.tsx [hunk 3]
```diff
@@ -6415,6 +6548,23 @@ function ChatViewContent(props: ChatViewProps) {
...
+      // from the draft goes now, before the resolve can bring the draft back
+      // on screen.
+      const carried = carriedComposerDraftByRequestIdRef.current.get(requestId);
+      if (carried) {
+        carriedComposerDraftByRequestIdRef.current.delete(requestId);
+        const draftPrompt =
+          useComposerDraftStore.getState().getComposerDraft(carried.draftTarget)?.prompt ?? "";
+        if (draftPrompt === carried.text) {
+          setComposerDraftPrompt(carried.draftTarget, "");
+        }
```
[the remark this answers is not in the document]
Comment (lnieuwenhuis) [reply apps/web/src/components/ChatView.tsx]: Fixed in a5826a768: on any failed or interrupted submit, the carried copy is put back into the draft store (and the carried entry restored) before the on-screen rescue runs, unless something else has filled the draft meanwhile. That path does not depend on which thread is on screen, so a reload or an off-screen cancel after a failed submit still finds the draft.

**Comment by lnieuwenhuis** [comment]
Release Smoke on a5826a768 fails with `ERR_PNPM_UNUSED_PATCH: expo-sharing@57.0.16`. It is unrelated to this PR: the smoke script deletes the lockfile and resolves fresh, `apps/mobile` asks for `~57.0.16`, and expo-sharing 57.0.17 is now on the registry, so the patch pinned to 57.0.16 goes unused. The same job failed identically on another PR run a few minutes earlier (run 33654817434) and passed on this branch's previous commit. It needs the patch bumped or the range pinned on main.

**Review by Mnigos — APPROVED** [review]
Retested after the latest commits and this is exactly what I was hoping for. Typed text now stays visible in the composer when the question card arrives, Stop still rescues everything into the draft, and a submitted answer doesn't come back duplicated after switching threads. Thanks for picking up the feedback so quickly.

File: apps/web/src/components/ChatView.tsx [hunk 4]
```diff
@@ -7225,6 +7375,23 @@ export default function ChatView(props: ChatViewProps) {
...
+      // on screen.
+      const carried = carriedComposerDraftByRequestIdRef.current.get(responseKey);
+      if (carried) {
+        carriedComposerDraftByRequestIdRef.current.delete(responseKey);
+        const draftPrompt =
+          useComposerDraftStore.getState().getComposerDraft(carried.draftTarget)?.prompt ?? "";
+        if (draftPrompt === carried.text) {
+          setComposerDraftPrompt(carried.draftTarget, "");
+        }
+      }
```
[the remark this answers is not in the document]
Comment (lnieuwenhuis) [reply apps/web/src/components/ChatView.tsx]: Fixed in 230009248. Submission bookkeeping now derives its key from composerDraftTarget consistently with carry/rescue; responseKey remains only for server-scoped in-flight deduplication. Verified the actual submission callback against a mounted draft route for success and interrupted failure (failed before the fix, passed afterward). 164 focused tests and web typecheck pass.

**Comment by shivamhwp** [comment]
Note: GPT-6 on behalf of shivam (@shivamhwp).

The failure handler still loses the edited answer when submission fails while its thread is offscreen. Start with a composer draft, carry it into a question, edit the answer, submit, switch to another thread, let the response fail there, and return after the question has disappeared. `beginSubmission` restores `carried.text`, which is the original draft; the question transition cannot rescue the edited answer because the previous visible thread was different.

Persist the captured submitted custom answers to their owning draft when the failure arrives, even if that owner is offscreen. Track any unrelated draft text separately from that persisted answer copy, so a later edit, erase, cancellation or successful retry replaces/removes only the answer and retains the unrelated text. Keep the environment-scoped request key throughout.

**Comment by lnieuwenhuis** [comment]
Fixed in b47fa2ffa. Failed submissions now persist the captured edited custom answer directly to the environment-scoped owning draft, including while another thread is visible. The persisted answer copy tracks unrelated draft text separately, so cancellation/edit replaces only the answer, erasure removes it, and a successful retry retains the unrelated draft.

Mounted regressions cover failure while offscreen followed by disappearance, edit, erase, and retry, plus identical thread/request IDs in another environment. Five assertions failed before the fix; all 158 focused hook/ChatView logic tests now pass. Web typecheck, targeted lint/format, and secrets scan pass. Also corrected the three test definite-assignment declarations from the previous CodeRabbit review.
